### PR TITLE
CBG-779: Add ActiveReplicator APIs and implement blipsync

### PR DESCRIPTION
**Pull request by bbrks**
- Moved some common BLIP code into new `replicator` package
- Added ActiveReplicator APIs/Config
- Perform initial blipsync via single node-loopback test

### CODE REVIEW

File: replicator/active_replicator_pull.go
Comment (adamcfraser) [inline replicator/active_replicator_pull.go]: Is there a reason to no-op instead of error if we call Connect() on a nil replicator?

File: replicator/active_replicator_pull.go
Comment (adamcfraser) [inline replicator/active_replicator_pull.go:57]: Since Close also closes the blipSender, it doesn't look like we support stopping/starting a replication while maintaining the connection.  Given that, what's the benefit of separate Connect() and Start() APIs?

**Review by adamcfraser — COMMENTED** [review]
Generally looks fine, but a couple of questions that I'd like feedback on before approving/merging.

File: replicator/active_replicator_pull.go
Comment (bbrks) [reply replicator/active_replicator_pull.go:57]: We should decide what happens when a one-shot has caught up. Should it disconnect, or keep the websocket open and wait for another Start?

Either sounds feasible, but disconnecting does sound safer to prevent a ton of open idle connections. If that is the case, I'll drop the connect to private, and roll it into Start.

File: replicator/active_replicator_pull.go
Comment (bbrks) [reply replicator/active_replicator_pull.go]: No reason, safer to error. Changed.

File: replicator/active_replicator_config.go
Comment (bbrks) [reply replicator/active_replicator_config.go]: Agreed, feels redundant. I've removed validation to simplify given the config will likely change to the existing ReplicatorConfig struct.

File: replicator/active_replicator.go
Comment (bbrks) [reply replicator/active_replicator.go:101]: Spent a bit of time investigating when writing this, and we have a couple of options:

- Continue as we do today, and ignore Origin on the passive side.
- Implement a config option for allowed origins on the passive side for SG-SG replications.
We can't really add this for CBL connections, because it's usually impossible to know what origins your clients are connecting from ahead of time.
Seems like this client validation for SG-to-SG replications would be better served by x509 auth.

For the first option, we still have to set a valid URL. It's non-trivial to get an accurate origin address without using something like `base.FindPrimaryAddr()`, which doesn't seem worth it if it's not being used on the passive side.

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**
